### ISSUE TITLE
Self-labels on Text Posts

### ISSUE THREAD

**Issue by foxyoreos** [bluesky-social/social-app#7011]
### Describe the Feature

(Apologies if there's already an issue for this, I searched but couldn't find one)

It makes no sense that self-labels on the web can only be applied to images. From what I can tell, self-labels *can* be applied to text posts but are blocked purely on the front-end.

I'm sure I'm oversimplifying, but this seems to be less about building a new feature and more about getting rid of code paths that turn off self-labeling or rendering warnings on text posts for signed out users.

Very specifically, what I'm suggesting is that the conditional rendering in the Composer view be removed so that the button shows up regardless of whether or not the post has media attached.

### Additional Context

I feel like I should write more here, but I don't know what else to say about it. Text can obviously be pornographic; I regularly post text on Bluesky that if I copied into this message as an example, would get me very quickly banned from Github. Particularly with the influx of furries and NSFW artists, horny text posts come along for the ride - and every concern about NSFW images apply to them. It's extremely important that minors on the network be able to filter them out. It would be embarrassing for them to show up unfiltered in public situations. And users just might find them gross, disturbing, or not want to read them.

I don't understand why the front-end would (as far as I can tell from the code) go out of its way to disable the feature without media. Adult content labels hide both images and text; the same UI that works for posts with images would also work for posts without images, unaltered. And from my own tests, Bluesky's APIs already happily accepts self-labels on text posts - it's possible I'm missing something, but about the only regression I can find is that adult labels don't get hidden in the UI from signed-out users. Which, I would hazard a guess, is also probably the result of a conditional rendering path in the front-end that could maybe be removed?

I'm a little bit baffled trying to come up with a reason why they might have been turned off in the UI originally. If the idea is that text can't really be pornographic in an upsetting way, then I invite anypony suggesting that to read through Bluesky's vore hashtag.

**Comment by MichaelGHUser** [comment]
Similar: https://github.com/bluesky-social/social-app/issues/1195

**Comment by foxyoreos** [comment]
Agree that the issues are similar (and I would like an "other" label as well); but I want to be 100% clear that what I am proposing here is not an expansion of the content warning system.

Pornographic labels can already be applied to text posts - the UI turns them off. I'm not asking for a new feature (although again, I absolutely support #1195). But in this specific issue, I'm asking for the UI to stop *disabling* a feature that already exists.

**Comment by foxyoreos** [comment]
Just to prevent this from auto-closing in case there's some kind of stale filter, reminder that it's still an issue and that yes, text posts can still be pornographic >w<

**Comment by foxyoreos** [comment]
In light of https://bsky.social/about/blog/07-10-2025-age-assurance I'm curious if we're going to actually get a way to label pornographic text we upload as pornographic.

The post ends with:

> In addition to these updates, we’ll continue to look at how we’re doing to shield children from unwanted content, and how we can improve.

But are you actually looking into it? Are you improving? There are so many proposals to make self-labeling better that have seen zero attention or even acknowledgement from the Bluesky developer team. This site lacks the most basic safety features imaginable. Stuff I post on Bluesky that would get my account *banned* on Github for pornographic content, and Bluesky's position is not only is it fine for kids to see that, but I shouldn't have a way as a user to even optionally stop kids from seeing it. Make it make sense.

Now Bluesky is going to start doing age verification on content that the website literally prevents from being labeled. How is that supposed to work? The site has put almost zero thought into NSFW labeling or safety features, and refuses to engage with NSFW communities and advocates to help make the site safer in any way.

This is a website that does not care about the safety of children or the safety and security of NSFW communities. And the fact that the site literally prevents pornographic content from being labeled says a lot about the actual intention and outcome of the site's evolution around age verification and NSFW material.

**Comment by foxyoreos** [comment]
It's 2026, and this issue still hasn't been addressed.

Self-labels are possible directly from the PDS but there's no UI for it. The site doesn't display NSFW labels properly to any other account viewing them. Bluesky *does* in fact have categories for labeling that would hide text behind warnings - these show up in multiple Bluesky built-in labels.

I sign into Bluesky and see another round of discourse about NSFW artists "shoving porn in people's faces." I look over at the this issue in the other open tab and sigh.

**Comment by MichaelGHUser** [comment]
@foxyoreos 

> This is a website that does not care about the safety of children...

Content labels should focus on user consent. Twitter, Bluesky, and Mastodon aren’t suitable for kids. At best, they’re appropriate only for _supervised_ older teens and adults. Parents should be monitoring their children; when that responsibility gets pushed onto platforms, governments eventually step in, as they are now, with draconian age-verification laws. I don’t want to live in that world.

**Comment by foxyoreos** [comment]
So I want to be super clear here, because I think you may be latching onto something unrelated to the main issue. I am asking for the ability to label *my own posts*. Not for Bluesky to label them. I want to be able to label my own posts as NSFW: the same way that I can label images, links, and videos.

This is a clearly useful feature - not just for kids, but for everyone who reads Bluesky in public, who may be sex averse, or who may just want warnings or a heads-up before they see pornographic threads.

That quote is a reference to [this comment](https://github.com/bluesky-social/social-app/issues/7011#issuecomment-3063566090), which I made in the context of Bluesky implementing age verification rules (but notably, Bluesky has *not* banned children from the platform or said that Bluesky is unsuitable for underage users). My comment was a response to Bluesky saying:

> we’ll continue to look at how we’re doing to shield children from unwanted content

And what I'm saying is that in a world where I can't label my own pornographic posts as NSFW, that sentence is meaningless and performative - Bluesky *doesn't* shield anyone from porn, it doesn't give me as a NSFW creator tools to put up even my own shields. If Bluesky is going to pretend that it is a safe platform for kids to be on (which it seems committed to pretending), I ought to be able to put a warning label on the hard-core kink that I post.

That's a very minor ask from me that is not an endorsement of age verification, or the platform policing porn, or anything like that. I want to be able to treat my text posts the same way I treat images.

I think I ought to be able to include a warning on this stuff even if there's not a single child on the app. It is baffling to me that Bluesky is somehow acting like labeling is a thing that only applies to images. And not just NSFW stuff - I can post the most graphic description of violence on Bluesky imaginable, and there's no mechanism for me to label my own stuff as graphic content. It's an indefensible oversight.

**Comment by MichaelGHUser** [comment]
You're right. I mistakenly latched onto the "protect the children" framing, probably just a knee-jerk reaction on my part. The core point stands: users should be able to self-label text posts just like media so that the audience can decide, with informed consent, whether they want to view that content or not.

**Comment by foxyoreos** [comment]
I probably should have phrased the original comment better - I didn't mean it as an endorsement of age verification as much as a reaction to Bluesky's own framing.

A better way I could have wrote it might have been something to the effect of: "why are we pretending that any of this is about protecting children, when I can't even label my posts as NSFW in the first place?"

But yep, the only scope I have to this issue is that I want to be able to do the same thing for text posts that I can do today for images. <3

**Comment by typeholes** [comment]
This goes beyond NSFW.  There are many conversations we should be able to have but can't currently have responsibly as they would be triggering to some people. Most would be NSFW, but they'd also be not safe for some people who are fine with pornography. Others might benefit from a supportive conversation on those topics.

**Comment by arimelody** [comment]
the fediverse has had user-assigned content warnings for text posts (and all other kinds of posts) for as long as i've been a member, and it's only been a blessing for myself and everyone else i've spoken to about it. frankly, it was an immediate downgrade i noted once i started actively posting on bluesky. the fact that bluesky devs have repeatedly ignored this issue is baffling, especially considering their push to implement region-based age verification as mentioned earlier in this issue.

you cannot seriously claim to care about safety of users if you aren't going to give them so much as the option to self-label their posts.

it's bad enough that bluesky's label system is bound to categories and not arbitrary text (imo, the lexicon could simply be expanded to a category + optional detail pair). the point of this thread is self-evident: text can be pornographic. text can have adult content. text can be suggestive, or contain graphic details (see: any news out of gaza the past N+1 years).

unless i'm missing some critical detail of the labels spec, this seems like a minor UI tweak. one that could greatly benefit user safety, and one that users have been asking at least the past two years for. sure, by that metric a third-party client could implement it, but social-app is ostensibly the flagship way most people interact with bluesky. third-party integrations at this time will only affect ~1% of users.
